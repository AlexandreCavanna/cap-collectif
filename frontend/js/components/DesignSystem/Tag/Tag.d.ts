// @ts-nocheck
import { AppBoxProps, PolymorphicComponent } from '../../Ui/Primitives/AppBox'
import { ComponentProps } from 'react'
import { ICON_NAME } from '../Icon/Icon'
import { AvatarProps } from '../Avatar/Avatar'

type Props = ComponentProps<'span'> & {
  readonly icon?: keyof typeof ICON_NAME
  readonly variantType?: 'tag' | 'badge'
  readonly avatar?: AvatarProps & { props?: AppBoxProps }
  readonly onRemove?: (e?: MouseEvent) => void
  readonly variant?: 'blue' | 'aqua' | 'red' | 'green' | 'orange' | 'yellow' | 'gray' | 'neutral-gray'
  readonly interactive?: boolean
}

declare const Tag: PolymorphicComponent<Props>

export default Tag
