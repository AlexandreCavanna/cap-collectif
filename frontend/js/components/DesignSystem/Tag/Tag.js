// @flow
import * as React from 'react';
import { useState } from 'react';
import { variant } from 'styled-system';
import styled from 'styled-components';
import { AnimatePresence, m as motion } from 'framer-motion';
import css from '@styled-system/css';
import type { AppBoxProps } from '~ui/Primitives/AppBox.type';
import AppBox from '~ui/Primitives/AppBox';
import Icon, { ICON_NAME } from '~ds/Icon/Icon';
import { ease } from '~/utils/motion';
import { Spacing } from '~ui/Primitives/constants';
import Text from '~ui/Primitives/Text';
import colors from '~/styles/modules/colors';
import jsxInnerText from '~/utils/jsxInnerText';
import type { AvatarProps } from '~ds/Avatar/Avatar';
import Avatar from '~ds/Avatar/Avatar';

export type TagVariant =
  | 'blue'
  | 'aqua'
  | 'red'
  | 'green'
  | 'orange'
  | 'yellow'
  | 'gray'
  | 'neutral-gray';

export type Props = {|
  ...AppBoxProps,
  +icon?: $Values<typeof ICON_NAME>,
  +variantType?: 'tag' | 'badge',
  +avatar?: {| ...AvatarProps, props?: AppBoxProps |},
  +onRemove?: (e?: MouseEvent) => void,
  +variant?: TagVariant,
  +interactive?: boolean,
|};

export const styles = {
  '& .icon': {
    color: 'inherit',
  },
};

const TagInner = styled(AppBox).attrs({
  maxHeight: Spacing.Xl,
  display: 'inline-flex',
  alignItems: 'center',
})(
  // $FlowFixMe flow types are wrong, it can also be a function which take the props and return an object
  ({ variantType, interactive }) => {
    return variant({
      variants: {
        blue: {
          bg: 'blue.150',
          '--current-shadow-color': colors.blue[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'blue.200' : undefined,
              }
            : {},
          color: 'blue.800',
        },
        aqua: {
          bg: 'aqua.150',
          '--current-shadow-color': colors.aqua[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'aqua.200' : undefined,
              }
            : {},
          color: 'aqua.800',
        },
        red: {
          bg: 'red.150',
          '--current-shadow-color': colors.red[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'red.200' : undefined,
              }
            : {},
          color: 'red.800',
        },
        green: {
          bg: 'green.150',
          '--current-shadow-color': colors.green[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'green.200' : undefined,
              }
            : {},
          color: 'green.800',
        },
        orange: {
          bg: 'orange.150',
          '--current-shadow-color': colors.orange[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'orange.200' : undefined,
              }
            : {},
          color: 'orange.800',
        },
        yellow: {
          bg: 'yellow.150',
          '--current-shadow-color': colors.yellow[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'yellow.200' : undefined,
              }
            : {},
          color: 'yellow.800',
        },
        gray: {
          bg: 'gray.150',
          '--current-shadow-color': colors.gray[200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'gray.200' : undefined,
              }
            : {},
          color: 'gray.800',
        },
        'neutral-gray': {
          bg: 'neutral-gray.150',
          '--current-shadow-color': colors['neutral-gray'][200],
          '&:hover': interactive
            ? {
                bg: variantType === 'tag' ? 'neutral-gray.200' : undefined,
              }
            : {},
          color: 'neutral-gray.800',
        },
      },
    });
  },
  variant({
    prop: 'variantType',
    variants: {
      tag: {
        px: 2,
        fontSize: 2,
        py: 1,
        fontWeight: 400,
        fontFamily: 'roboto',
        maxWidth: '150px',
      },
      badge: {
        px: 4,
        fontSize: 1,
        py: 2,
        fontWeight: 600,
        fontFamily: 'openSans',
        textTransform: 'uppercase',
      },
    },
  }),
);

const IconContainer = motion(AppBox);

const Tag = React.forwardRef<Props, HTMLSpanElement>(
  (
    { children, icon, onRemove, avatar, variantType = 'tag', interactive = true, ...props }: Props,
    ref,
  ) => {
    const [state, set] = useState<'idle' | 'hovering'>('idle');

    const onFocus = () => {
      if (!onRemove) return;
      if (interactive) set('hovering');
    };

    return (
      <TagInner
        ref={ref}
        css={styles}
        variantType={variantType}
        title={jsxInnerText(children)}
        position="relative"
        borderRadius="tags"
        {...props}
        onMouseLeave={() => {
          if (interactive) set('idle');
        }}
        onMouseEnter={onFocus}
        onClick={onFocus}
        {...props}>
        {avatar && (avatar.src || avatar.name) && (
          <Avatar
            {...avatar.props}
            size="xs"
            mr={1}
            src={avatar.src}
            name={avatar.name}
            alt={avatar.alt}
          />
        )}
        {icon && <Icon name={icon} size="sm" mr={1} />}
        <Text
          lineHeight="sm"
          fontFamily="inherit"
          css={css({
            overflow: 'hidden',
            whiteSpace: 'nowrap',
            paddingRight: state === 'hovering' ? (variantType === 'tag' ? '15%' : '1%') : undefined,
            textOverflow: 'ellipsis',
          })}>
          {children}
        </Text>
        {onRemove && (
          <AnimatePresence>
            {state === 'hovering' && (
              <IconContainer
                onClick={onRemove}
                ml={1}
                css={{ '&:hover': { cursor: 'pointer' } }}
                display="flex"
                bg="inherit"
                boxShadow="-5px 0 2px var(--current-shadow-color)"
                initial={{ opacity: 0, position: 'absolute', right: 0 }}
                animate={{ opacity: 1, position: 'absolute', right: Spacing.Xs2 }}
                transition={{ duration: 0.2, ease }}
                exit={{ opacity: 0, position: 'absolute', right: 0 }}>
                <Icon className="cross-icon" name={ICON_NAME.CROSS} color="inherit" size="xs" />
              </IconContainer>
            )}
          </AnimatePresence>
        )}
      </TagInner>
    );
  },
);

Tag.displayName = 'Tag';

export default Tag;
