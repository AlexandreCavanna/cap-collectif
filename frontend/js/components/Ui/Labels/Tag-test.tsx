/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { Tag } from './Tag'

const children = 'Label'
describe('<Tag />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Tag id="example">{children}</Tag>)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render an icon if defined', () => {
    const wrapper = shallow(<Tag icon="cap cap-icon">{children}</Tag>)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render with different html tag', () => {
    const wrapper = shallow(
      <Tag as="span" icon="cap cap-icon">
        {children}
      </Tag>,
    )
    expect(wrapper).toMatchSnapshot()
  })
  it('should render a specific container for link', () => {
    const wrapper = shallow(
      <Tag as="span" href="#" icon="cap cap-icon">
        {children}
      </Tag>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
