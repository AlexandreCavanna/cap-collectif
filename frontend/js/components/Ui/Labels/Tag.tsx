import * as React from 'react'
import type { StyledComponent } from 'styled-components'
import styled from 'styled-components'

export const TagContainer: StyledComponent<any, {}, HTMLSpanElement> = styled.span.attrs({
  className: 'tag',
})`
  font-size: 14px;
  display: inline-block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
  border: 0;
  margin: 0;
  padding: 0 0 0 5px;
  color: inherit;

  &:focus,
  &:hover {
    text-decoration: none;
  }

  & > .icon,
  & > .icon-rounded {
    margin-right: 5px;
  }

  .customImage {
    margin: 0;
    padding: 0 5px 0 0;
    vertical-align: top;

    * {
      margin: 0;
      padding: 0;
    }
  }
`
const TagLink: StyledComponent<any, {}, HTMLSpanElement> = styled(TagContainer)`
  &:hover {
    cursor: pointer;
  }
`
const TagIcon = styled.i<{ size: any }>`
  padding-right: 5px;
  font-size: ${props => props.size || '14px'};
`
type Props = {
  children: any
  size?: string
  as?: any
  icon?: string
  className?: string
  CustomImage?: any
  onClick?: () => void
  id?: string
  href?: string
  bsStyle?: string
  rel?: string
  target?: string
  title?: string
}
export class Tag extends React.Component<Props> {
  render() {
    const { size, children, as, icon, CustomImage, className, ...rest } = this.props
    const Container = as === 'a' ? TagLink : TagContainer
    return (
      <Container className={`tag ${className || ''}`} as={as} {...rest}>
        {icon && <TagIcon size={size} className={icon} />}
        {CustomImage &&
          React.cloneElement(CustomImage, {
            className: 'customImage ',
          })}
        {children}
      </Container>
    )
  }
}
export default Tag
