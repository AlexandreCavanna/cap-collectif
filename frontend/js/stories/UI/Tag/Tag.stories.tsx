// @ts-nocheck
import * as React from 'react'
import Tag from '~ds/Tag/Tag'
import Flex from '~ui/Primitives/Layout/Flex'
import { ICON_NAME } from '~ds/Icon/Icon'

export default {
  title: 'Design system/Tag',
  component: Tag,
  argTypes: {
    icon: {
      control: {
        type: 'select',
        options: Object.values(ICON_NAME),
      },
    },
  },
}

const Template = (args: any) => (
  <>
    <Flex gridGap={2} wrap="wrap">
      <Tag variant="aqua">Label</Tag>
      <Tag variant="blue">Et moi un label encore plus long</Tag>
      <Tag variant="gray">Bonsoir comment</Tag>
      <Tag variant="green">allez vous en</Tag>
      <Tag variant="neutral-gray">ce</Tag>
      <Tag variant="orange">doux</Tag>
      <Tag variant="red">confinement</Tag>
      <Tag variant="yellow">Pls like share & subscribe</Tag>
    </Flex>
    <Flex gridGap={2} wrap="wrap">
      <Tag {...args}>Use controls to change this one</Tag>
    </Flex>
  </>
)

export const main = Template.bind({})
main.storyName = 'Default'
main.args = {}

const IconTemplate = () => (
  <Flex gridGap={2} wrap="wrap">
    <Tag icon={ICON_NAME.ARROW_UP} variant="neutral-gray">
      Go Up
    </Tag>
    <Tag icon={ICON_NAME.ARROW_RIGHT} variant="green">
      & Right
    </Tag>
    <Tag icon={ICON_NAME.ARROW_DOWN} variant="aqua">
      & Down
    </Tag>
    <Tag icon={ICON_NAME.ARROW_LEFT} variant="gray">
      & Left
    </Tag>
    <Tag icon={ICON_NAME.ADD} variant="blue">
      0 + 0 = la tête à toto ^^ :p
    </Tag>
    <Tag icon={ICON_NAME.BELL} variant="blue">
      Dring dring
    </Tag>
    <Tag icon={ICON_NAME.CLOCK} variant="orange">
      Confined
    </Tag>
    <Tag icon={ICON_NAME.TRASH} variant="red">
      Benett est trash tier
    </Tag>
    <Tag icon={ICON_NAME.PENCIL} variant="yellow">
      Dessin
    </Tag>
  </Flex>
)

export const withIcons = IconTemplate.bind({})
withIcons.storyName = 'with icons'
withIcons.args = {}

const AvatarTemplate = () => (
  <Flex gridGap={2} wrap="wrap">
    <Tag
      avatar={{
        src: 'https://risibank.fr/cache/stickers/d1261/126102-full.png',
        name: 'Mikasa Estucasa',
      }}
      variant="neutral-gray"
    >
      Mikasa Estucasa
    </Tag>
    <Tag
      avatar={{
        name: 'Dan Abramov',
        src: 'https://bit.ly/dan-abramov',
      }}
      variant="yellow"
    >
      Dan Abramov
    </Tag>
    <Tag
      avatar={{
        name: 'Omar Jbara',
        props: {
          bg: 'blue.500',
          color: 'white',
        },
      }}
      variant="blue"
    >
      Omar Jbara
    </Tag>
  </Flex>
)

export const withAvatars = AvatarTemplate.bind({})
withAvatars.storyName = 'with Avatars'
withAvatars.args = {}

const RemoveActionTemplate = () => (
  <Flex gridGap={2} wrap="wrap">
    <Tag
      onRemove={() => {
        window.alert("J'espère que tu as rigolé :p :) ~.~")
      }}
      icon={ICON_NAME.ADD}
      variant="blue"
    >
      0 + 0 = toto
    </Tag>
    <Tag
      onRemove={() => {
        window.alert('Coucou')
      }}
      variant="neutral-gray"
    >
      Bonsoir à tous et à toutes
    </Tag>
    <Tag
      onRemove={() => {
        window.alert('Tik Tok, on the clock')
      }}
      icon={ICON_NAME.CLOCK}
      variant="orange"
    >
      Confined
    </Tag>
    <Tag
      onRemove={() => {
        window.alert('La vérité blesse, désolée...')
      }}
      icon={ICON_NAME.TRASH}
      variant="red"
    >
      Benett trash
    </Tag>
    <Tag
      onRemove={() => {
        window.alert('pas cool dis donc')
      }}
      icon={ICON_NAME.CIRCLE_CROSS}
      variant="red"
    >
      Salut
    </Tag>
    <Tag
      avatar={{
        name: 'Erwin Smith',
        src: 'https://pm1.narvii.com/6981/6dd582abb1862f56713069173d2c2fbe9b9f68bar1-300-300v2_hq.jpg',
      }}
      onRemove={() => {
        window.alert('cool dis donc')
      }}
      variant="green"
    >
      Supprime moi
    </Tag>
  </Flex>
)

export const withonRemove = RemoveActionTemplate.bind({})
withonRemove.storyName = 'with onRemove actions'
withonRemove.args = {}

const BadegeTemplate = () => (
  <Flex align="center" gridGap={2} wrap="wrap">
    <Tag variantType="badge" variant="aqua">
      Label
    </Tag>
    <Tag icon="CLOCK" variantType="badge" variant="blue">
      Débute dans 21 jours
    </Tag>
    <Tag icon="CIRCLE_CROSS" variantType="badge" variant="red">
      Confined
    </Tag>
    <Tag variantType="badge" icon="BELL" variant="yellow">
      Pls like share & subscribe
    </Tag>
  </Flex>
)

export const asBadge = BadegeTemplate.bind({})
asBadge.storyName = 'as Badge'
asBadge.args = {}
