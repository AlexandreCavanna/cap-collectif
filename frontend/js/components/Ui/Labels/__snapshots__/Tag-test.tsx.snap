// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> should render a specific container for link 1`] = `
<Tag__TagContainer
  as="span"
  className="tag "
  href="#"
>
  <Tag__TagIcon
    className="cap cap-icon"
  />
  Label
</Tag__TagContainer>
`;

exports[`<Tag /> should render an icon if defined 1`] = `
<Tag__TagContainer
  className="tag "
>
  <Tag__TagIcon
    className="cap cap-icon"
  />
  Label
</Tag__TagContainer>
`;

exports[`<Tag /> should render correctly 1`] = `
<Tag__TagContainer
  className="tag "
  id="example"
>
  Label
</Tag__TagContainer>
`;

exports[`<Tag /> should render with different html tag 1`] = `
<Tag__TagContainer
  as="span"
  className="tag "
>
  <Tag__TagIcon
    className="cap cap-icon"
  />
  Label
</Tag__TagContainer>
`;
